### PR TITLE
Special Orthogonal Group

### PR DESCRIPTION
**Pull request by stoprightthere**
PR introduces a new space, Special Orthogonal Group SO(n).

### CODE REVIEW

File: geometric_kernels/lab_extras/extras.py [hunk 1]
```diff
@@ -11,6 +11,26 @@
...
+@dispatch
+@abstract()
+def take_along_last_axis(a: B.Numeric, indices: B.Numeric) -> B.Numeric:
+    """
+    Takes elements of `a` along the last axis. `indices` must be the same rank (ndim) as
+    `a`. Useful in e.g. argsorting and then recovering the sorted array.
+    ```
+
+    E.g. for 3d case:
+    output[i, j, k] = a[i, j, indices[i, j, k]]
```
Comment (vdutor) [inline geometric_kernels/lab_extras/extras.py:33]: Why does `indices` need to have the same rank as `a`? Would this not work?
```suggestion
    output[i, j, k] = a[i, j, indices[i, j]]
```

File: geometric_kernels/lab_extras/extras.py [hunk 2]
```diff
@@ -20,3 +40,29 @@ def from_numpy(_: B.Numeric, b: Union[List, B.Numeric, B.NPNumeric]):
...
+@abstract()
+def copysign(a: B.Numeric, b: B.Numeric) -> B.Numeric:
+    """
+    Change the sign of `a` to that of `b`, element-wise.
+    """
+
+
+def isclose(a: B.Numeric, b: B.Numeric, rtol=1e-5, atol=1e-8):
+    """Returns a boolean array where two arrays are element-wise equal within a
+    tolerance."""
```
Comment (vdutor) [inline geometric_kernels/lab_extras/extras.py:63]: Minor: consistency in docs formatting
```suggestion
    """
    Returns a boolean array where two arrays are element-wise equal within a
    tolerance.
    """
```

File: tests/spaces/test_so.py [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+
+    kernel = MaternKarhunenLoeveKernel(so_group, 16)
+    params, state = kernel.init_params_and_state()
+    test_so_group((kernel, params, state), size=3)
+
+    test_so_kernel_torch((kernel, params, state))
+
+    test_so_kernel_tf((kernel, params, state))
+
+    test_so_kernel_jax((kernel, params, state))
```
Comment (vdutor) [inline tests/spaces/test_so.py:112]:
```suggestion
```

File: tests/spaces/test_so.py [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+
+# parametrize by size
+@pytest.mark.parametrize("size", [5, 6, 7, 8, 9, 10])
+def test_so_group(so_kernel, size):
+    kernel, params, state = so_kernel
+
+    dim = kernel.space.dimension
+    rank = dim // 2
+    so_dim = 2 * rank ** 2 + rank if dim % 2 else 2 * rank ** 2 - rank
+    print(f"\ntesting dim={dim} (dimSO={so_dim}, rank={rank})")
```
Comment (vdutor) [inline tests/spaces/test_so.py:47]:
```suggestion
```

File: tests/spaces/test_so.py [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+    m = random_so_matrix(dim, size)
+
+    K = kernel.K(params, state, m, m)
+
+    assert K.shape == (size, size)
+
+    np.linalg.cholesky(K + 1e-6 * np.eye(K.shape[0]))
+
+
+def test_so_kernel_torch(so_kernel):
```
Comment (vdutor) [inline tests/spaces/test_so.py:58]: Given the similarity between `test_so_kernel_torch`, `test_so_kernel_tensorflow` and `test_so_kernel_jax`, we can make this one test function and parameterize the backend, like I did for the tests in `test_circle`: https://github.com/GPflow/GeometricKernels/blob/main/tests/spaces/test_circle.py#L34

File: geometric_kernels/spaces/special_orthogonal_group.py [hunk 6]
```diff
@@ -0,0 +1,408 @@
...
+    def get_eigenfunctions(self, num):
+        """
+        :param num: number of eigenfunctions returned.
+        """
+        if num not in self.cache:
+            eigenfunctions = SOEigenfunctions(self.dim, num)
+            self.cache[num] = eigenfunctions
+            return eigenfunctions
+        else:
+            return self.cache[num]
```
Comment (vdutor) [inline geometric_kernels/spaces/special_orthogonal_group.py:408]: can we use lru cache for this: https://docs.python.org/3/library/functools.html?

File: geometric_kernels/spaces/special_orthogonal_group.py [hunk 7]
```diff
@@ -0,0 +1,408 @@
...
+        :param n: Dimensionality `SO(n)`, aka the size of the matrices.
+        """
+        self.cache: Dict[int, Any] = {}
+        super().__init__(n=n)
+
+    @property
+    def dimension(self):
+        return self.n
+
+    def get_eigenvalues(self, num: int):
```
Comment (vdutor) [inline geometric_kernels/spaces/special_orthogonal_group.py:381]: Can we use lru cache for this: https://docs.python.org/3/library/functools.html

File: geometric_kernels/spaces/special_orthogonal_group.py [hunk 8]
```diff
@@ -0,0 +1,408 @@
+"""
+Special orthogonal group, SO(n).
```
Comment (vdutor) [inline geometric_kernels/spaces/special_orthogonal_group.py:2]: Code in this file is very hard to properly review. Is there a paper we can link and match some of the equations in there to bits of the code?

File: geometric_kernels/lab_extras/extras.py (hunk 1 above)
Comment (stoprightthere) [reply geometric_kernels/lab_extras/extras.py:33]: This solution is the only one I found that works for all `np`, `tf`, `torch` and `jax`. E.g `torch.gather` requires indices be the same rank as the input array, while `tf.gather` uses a completely different `indices` structure. However, when indices are the same rank as the input, **and** the elements are taken along the last axis, this works with every framework.

Prob should write a unit-test to double-check that.

File: geometric_kernels/spaces/special_orthogonal_group.py (hunk 8 above)
Comment (stoprightthere) [reply geometric_kernels/spaces/special_orthogonal_group.py:2]: I agree, the code is hard to read. Unfortunately there's no publicly accessible paper to check against

File: geometric_kernels/kernels/geometric_kernels.py [hunk 9]
```diff
@@ -59,7 +59,7 @@ def init_params_and_state(self):
 
         :return: tuple(params, state)
         """
-        params = dict(lengthscale=1.0, nu=0.5)
+        params = dict(lengthscale=np.r_[1.0], nu=np.r_[0.5])
```
Comment (MJHutchinson) [inline geometric_kernels/kernels/geometric_kernels.py:62]: Should we be using np directly? Is there a LAB call to make array instead?

File: geometric_kernels/spaces/special_orthogonal_group.py [hunk 10]
```diff
@@ -0,0 +1,408 @@
...
+                    1,
+                )
+                return int(round(rep_dim))
+            else:
+                qs = [
+                    pk + self.rank - k - 1 if k != self.rank - 1 else abs(pk)
+                    for k, pk in enumerate(signature)
+                ]
+                rep_dim = int(
+                    reduce(
```
Comment (MJHutchinson) [inline geometric_kernels/spaces/special_orthogonal_group.py:325]: Is this going to be performant?

Comment (MJHutchinson) [reply geometric_kernels/spaces/special_orthogonal_group.py:325]: Or would it be faster to use LAB functions to do this?

File: geometric_kernels/spaces/special_orthogonal_group.py [hunk 11]
```diff
@@ -0,0 +1,408 @@
...
+        partition[j] = x = partition[j] + 1
+        j -= 1
+        while j > 0:
+            partition[j] = x
+            s -= x
+            j -= 1
+        partition[0] = s
+
+
+class SOEigenfunctions(Eigenfunctions):
```
Comment (MJHutchinson) [inline geometric_kernels/spaces/special_orthogonal_group.py:73]: Is there no addition theorem for the SO case?

File: geometric_kernels/kernels/geometric_kernels.py (hunk 9 above)
Comment (stoprightthere) [reply geometric_kernels/kernels/geometric_kernels.py:62]: It's okay, lab is essentially a wrapper and everything could be converted from numpy.

File: tests/spaces/test_so.py [hunk 12]
```diff
@@ -0,0 +1,112 @@
+import jax.numpy as jnp
```
Comment (NoemieJaquier) [inline tests/spaces/test_so.py:1]: As mentioned yesterday, we can use this code: https://github.com/NoemieJaquier/MaternGaBO/blob/main/BoManifolds/kernel_utils/kernels_so.py as "unofficial" test, if needed.

File: geometric_kernels/spaces/special_orthogonal_group.py [hunk 13]
```diff
@@ -0,0 +1,408 @@
...
+    """Eigenfunctions for Special Orthogonal Group SO(n)."""
+
+    def __init__(self, dim: int, num_representations: int = 10):
+        r"""
+        :param dim: dimensionality in which the group acts.
+        :param num_representation: number of representations of the group to compute. Values larger than 10 are impractical.
+        """
+        assert (
+            dim != 4
+        ), "Dimension 4 is not supported since SO(4) is not a semisimple group."
```
Comment (NoemieJaquier) [inline geometric_kernels/spaces/special_orthogonal_group.py:83]: Is this an issue for some applications?

File: tests/spaces/test_so.py (hunk 5 above)
Comment (stoprightthere) [reply tests/spaces/test_so.py]: Fixed

File: geometric_kernels/spaces/special_orthogonal_group.py (hunk 10 above)
Comment (stoprightthere) [reply geometric_kernels/spaces/special_orthogonal_group.py:325]: It does not take long to compute (for a reasonable number of eigfs). Could be possible to rewrite it in numpy (LAB per se does not induce any performance gains) but I don't know if it worth it

File: geometric_kernels/spaces/special_orthogonal_group.py [hunk 14]
```diff
@@ -0,0 +1,408 @@
...
+    r"""
+    Special orthogonal group `SO(n)`, that is a group of orthogonal matrices
+    with determinant equal to 1.
+    """
+
+    def __init__(self, n: int):
+        r"""
+        :param n: Dimensionality `SO(n)`, aka the size of the matrices.
+        """
+        self.cache: Dict[int, Any] = {}
```
[the remark this answers is not in the document]
Comment (stoprightthere) [reply geometric_kernels/spaces/special_orthogonal_group.py:374]: Same as above, do not think it will provide any performance gains.

File: geometric_kernels/spaces/special_orthogonal_group.py (hunk 11 above)
Comment (stoprightthere) [reply geometric_kernels/spaces/special_orthogonal_group.py:73]: Nope.

File: geometric_kernels/spaces/special_orthogonal_group.py (hunk 13 above)
Comment (stoprightthere) [reply geometric_kernels/spaces/special_orthogonal_group.py:83]: Solely sanity checks, for the dimension 4 the math does not work.

**Review by vdutor — APPROVED** [review]

**Comment by lrozo** [comment]
@stoprightthere , I was about to review this PR but it seems it has been thoroughly reviewed by several people already and you already addressed several fixes. Is there anything stopping you from merging to master?

**Comment by vabor112** [comment]
See PR #83
